lib/msal-browser/src/app/ClientApplication.ts

```diff
@@ -299,6 +308,64 @@ export abstract class ClientApplication {
+     * This authorization code should be acquired server-side using a confidential client to acquire a spa_code.
+     * This API is not indended for normal authorization code acquisition and redemption.
+     * 
+     * Redemption of this authorization code will not require PKCE, as it was acquired by a confidential client.
+     * 
+     * @param request {@link AuthorizationCodeRequest}
+     * @returns A promise that is fulfilled when this function has completed, or rejected if an error was raised.
+     */
+    async acquireTokenByCode(request: AuthorizationCodeRequest): Promise<AuthenticationResult> {
+        this.preflightBrowserEnvironmentCheck(InteractionType.Silent);
+        this.logger.trace("acquireTokenByCode called", request.correlationId);
+        this.eventHandler.emitEvent(EventType.ACQUIRE_TOKEN_BY_CODE_START, InteractionType.Silent, request);
+
+        try {
+            if (!request.code) {
+                throw BrowserAuthError.createAuthCodeRequiredError();
+            }
+
+            let response = this.hybridAuthCodeResponses.get(request.code);
+            if (!response) {
+                this.logger.verbose("Initiating new acquireTokenByCode request", request.correlationId);
+                response = this.acquireTokenByCodeAsync(request)
+                    .then((result: AuthenticationResult) => {
+                        this.hybridAuthCodeResponses.delete(request.code);
+                        return result;
+                    })
+                    .catch((error: Error) => {
+                        this.hybridAuthCodeResponses.delete(request.code);
+                        throw error;
+                    });
+
+                this.hybridAuthCodeResponses.set(request.code, response);
+            } else {
+                this.logger.verbose("Existing acquireTokenByCode request found", request.correlationId);
+            }
+            
+            return response;
+        } catch (e) {
+            this.eventHandler.emitEvent(EventType.ACQUIRE_TOKEN_BY_CODE_FAILURE, InteractionType.Silent, null, e);
+            throw e;
+        }
+    }
+
+    /**
+     * Creates a SilentAuthCodeClient to redeem an authorization code.
+     * @param request 
+     * @returns Result of the operation to redeem the authorization code
+     */
+    private async acquireTokenByCodeAsync(request: AuthorizationCodeRequest): Promise<AuthenticationResult> {
+        this.logger.trace("acquireTokenByCodeAsync called", request.correlationId);
+        const silentAuthCodeClient = new SilentAuthCodeClient(this.config, this.browserStorage, this.browserCrypto, this.logger, this.eventHandler, this.navigationClient, ApiId.acquireTokenByCode, request.correlationId);
+        const silentTokenResult = await silentAuthCodeClient.acquireToken(request);
+        this.eventHandler.emitEvent(EventType.ACQUIRE_TOKEN_BY_CODE_SUCCESS, InteractionType.Silent, silentTokenResult);
```

Comment:
Can we put this in the `.then` above so we have both success and failure events in the same place?

Reply:
Yes, good catch.